### PR TITLE
Add BaseSubscriber to write subscribers with request

### PR DESCRIPTION
**Pull request by simonbasle**
Name and location to be discussed.

cc @smaldini @mp911de

### CODE REVIEW

**Comment by smaldini** [comment]
A problem i can see is that its in publisher package, although it might be just fine.

File: src/main/java/reactor/core/publisher/BaseSubscriber.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+		return subscription != null;
+	}
+
+	/**
+	 * Hook for further processing of onSubscribe's Subscription.
+	 * Defaults to requesting {@code Long.MAX_VALUE}.
+	 *
+	 * @param subscription the subscription to optionally process
+	 */
+	protected void doOnSubscribe(Subscription subscription) {
```
Comment (simonbasle) [inline src/main/java/reactor/core/publisher/BaseSubscriber.java]: the prefix for the hooks that the user is supposed to override being `doOn` is probably confusing...

**Comment by artembilan** [comment]
Looks like I need this as well instead of my home-cooky: https://github.com/spring-projects/spring-integration/blob/master/spring-integration-core/src/main/java/org/springframework/integration/endpoint/ReactiveConsumer.java#L198

**Comment by simonbasle** [comment]
@smaldini what about pushing this into a `subscriber` package?

The `doOn...` methods should probably be renamed with a different prefix.

For the class name, I was thinking either

- [ ] `AbstractSubscriber`
- [ ] `RequestingSubscriber`
- [ ] but we can keep `BaseSubscriber`

**Comment by mp911de** [comment]
+1 for `BaseSubscriber`. [`do…`](https://github.com/spring-projects/spring-framework/blob/master/spring-jdbc/src/main/java/org/springframework/jdbc/core/JdbcTemplate.java#L434) method naming is often used in Spring Framework but it's not necessarily an indicator for [overridable hooks](https://github.com/spring-projects/spring-framework/blob/master/spring-jdbc/src/main/java/org/springframework/jdbc/core/JdbcTemplate.java#L1312).

**Comment by simonbasle** [comment]
rebased into a single commit, ready for squash/rebase-and-merge

**Comment by sdeleuze** [comment]
Feature wanted on my side, we may have some use case in the NIO Multipart Reactive implementation, and I think this is a good idea to provide an effective way to implement properly a `Subscriber`. I have no better name to propose, and using `publisher` package looks a little bit strange but given current package arrangement, I think we have no other choice.

But curious to know your thoughts about where this class will end up in Reactor 3.1? Maybe keep current structure and just use `reactor.core` instead of `reactor.core.publisher` given the fact that avoiding package tangle is hard with more fine grained structure?

**Comment by smaldini** [comment]
I thought about SimpleSubscriber its mostly a support class for our publishers our Publisher in general anyway. There is no meaning in Subscriber alone.

**Comment by sdeleuze** [comment]
If `BaseSubscriber` is not an abstract class, `SimpleSubscriber` is perhaps a better name indeed + it describes quite well what is the class.

**Comment by simonbasle** [comment]
it is *conceptually* abstract, not doing much by itself if you don't override the hooks, but all hooks have a default implementation... maybe `hookOnNext` should be abstract, as it currently is no-op and is probably the one users want to override the most?

**Comment by smaldini** [comment]
But SimpleSubscriber alone works for cases like .subscribe().

**Comment by simonbasle** [comment]
true, however I think we should encourage its usage in place of `.subscribe()`, no should we encourage its usage over the lambda variants of `.subscribe(v -> ...)` (if there is no need for `request`)

**Comment by sdeleuze** [comment]
People will may be confused if they can use this class as a regular class they instantiate or an SPI they should extends. Maybe better to start with an `abstract` class called `BaseSubscriber`, and optionally later add a non-abstract `SimpleSubscriber` class if there is a real need for that?

**Comment by simonbasle** [comment]
how about making it explicitly `abstract` so that it cannot be instantiated as is, clearly communicating this is a SPI, but still having all the `hookOnXXX` methods concrete?

**Comment by sdeleuze** [comment]
Is the goal is to avoid for the users to implement all the hooks if he just need one? What will be the most common use case from your POV?

**Comment by simonbasle** [comment]
I think the most common use case would be to override `hookOnSubscribe` and `hookOnNext` to perform custom-sized `request(n)` calls.

**Comment by sdeleuze** [comment]
I think I would declare `hookOnSubscribe` and `hookOnNext` as abstract, and let others as they are (complete and cancel no-op and error thrown by default with Javadoc explaining what default impl does) with some hints in the Javadoc about the typical use cases. That seems to be a good tradeoff.

**Pull request merged**